graphium/ipu/ipu_wrapper.py

```diff
@@ -101,7 +101,7 @@ def compute_loss(
 
     def on_train_batch_end(self, outputs, batch, batch_idx):
         outputs = self.convert_from_fp16(outputs)
-        outputs["loss"] = outputs["loss"].mean()
+        outputs["loss"] = outputs["loss"][outputs["loss"] != 0].mean()
```

Comment:
I'm really not sure about this change. This will cause an issue if the loss is actually 0 (which is very unlikely).

But what I'm most concerned about is that this only changes the loss that gets logged, not the gradient accumulation. 

Here's an example of how gradient accumulation is done in pytorch:

```python
optimizer.zero_grad()
for i, (inputs, labels) in enumerate(train_loader):
    outputs = model(inputs)
    loss = criterion(outputs, labels)
    loss = loss / accumulation_steps
    loss.backward()
    if (i + 1) % accumulation_steps == 0:
        optimizer.step()
        optimizer.zero_grad()
```

You can see that the loss gets divided by the `accumulation_steps` early on, but it's only the `backward` passes that gets accumulated later. If, after you're done with your steps, but before the call to `optimizer.step()`, you decide to rescale the loss, then it's too late.

Reply:
Instead, the better way of doing this is:
1. For all tasks, check if `labels.isnan().all()`, instead of checking if the loss is `0`
2. If `True` for `task_K`, then you can detach the task head `self.model.task_heads[task_K].requires_grad = False`

But because of the `if` statements, I don't think this is actually possible when compiled on IPU's.

When we discussed last time to check if 0-loss meant 0-gradient, I think the answer is yes since it means the model is perfect.

The issue then comes from the division by `accumulation_steps`.

Reply:
The loss would not be actually 0 because we do gradient descent on summation of 5 task losses? The 0 loss would be mostly on the mcf7 and vcap I think?